R/util.R

```diff
@@ -142,3 +142,75 @@ lookupByName = function(listpath, name, prop = "name") {
   assign(key, val, envir = byNameHash)
   val
 }
+
+allResources <- c("experiments", "attachments", "compensations", "events", "fcsfiles", "gates", "populations", "scalesets", "statistics")
+fuzzyMatch = function(string) {
+  return(agrep(pattern=string, allResources, value=T))
+
+getResourcesByName = function(resource, name){
+  all_ids = getExperiments()["_id"]
+  getter = getGenericResource(resource)
+  results = list()
+  for(row in 1:nrow(all_ids)){
+    id = all_ids[row,]
+    listpath = sprintf("experiments/%s/%s", id, resource)
+    # remove caches values from byNameHash environ in case of duplicate names
+    # this isn't a great way to deal with this, pending review
+    key = paste(listpath, resource, sep="$$$")
+    if (exists(key, envir = byNameHash)) rm(key, envir = byNameHash)
+
+    result = try(getter(id, byName(name)), silent=TRUE)
+    if(class(result) != "try-error") {
+      results = rbind(results, result)
+    }
+  }
+  return(results)
+}
+
+}
+
+getGenericResource = function(resource) {
+  getterList = list("experiments" = getExperiment,
+                    #"attachments" = getAttachment,
+                    #"compensations" = getCompensation,
+                    #"events" = getEvent,
+                    "fcsfiles" = getFcsFile,
+                    "gates" = getGate,
+                    "populations" = getPopulation)
+                    #"scalesets" = getScaleSet,
+                    #"statistics" = getStatistics)
+  func = getterList[[resource]]
+  if (is.null(func)) stop("No resource found with that name.")
+  return(func)
+}
+
+#' Find a Resource By Name
+#'
+#' Searches for any resource by name. Accepts imperfect resource
+#' specification, (i.e. "gate" for "gates").
+#'
+#' Because two resources in different experiments could have the
+#' same name, this clears the cache of ids and names each time it runs.
+#' It is recommended that you use the regular lookup functions
+#' for performant code.
```

Comment:
This function should take the `experimentId`, otherwise it'll be an impossibly large search space. That avoids this issue.

Reply:
It seemed from the issue description that we wanted to be able to query _any_ name of any resource, which requires checking all experiments; do you want it to work like `findByName(experimentId, resourceType, name)` instead?